crates/cheatcodes/src/test/assert.rs

```diff
@@ -0,0 +1,1011 @@
+    Lt { left: &'a T, right: &'a T },
+}
+
+macro_rules! format_values {
+    ($self:expr, $format_fn:expr) => {
+        match $self {
+            Self::NotEq { left, right } => format!("{} == {}", $format_fn(left), $format_fn(right)),
+            Self::Eq { left, right } => format!("{} != {}", $format_fn(left), $format_fn(right)),
+            Self::Ge { left, right } => format!("{} < {}", $format_fn(left), $format_fn(right)),
+            Self::Gt { left, right } => format!("{} <= {}", $format_fn(left), $format_fn(right)),
+            Self::Le { left, right } => format!("{} > {}", $format_fn(left), $format_fn(right)),
+            Self::Lt { left, right } => format!("{} >= {}", $format_fn(left), $format_fn(right)),
+        }
+    };
+}
+
+impl<'a, T: Display> ComparisonAssertionError<'a, T> {
+    fn format_for_values(&self) -> String {
+        format_values!(self, T::to_string)
+    }
+}
+
+#[allow(clippy::redundant_closure_call)]
+impl<'a, T: Display> ComparisonAssertionError<'a, Vec<T>> {
+    fn format_for_arrays(&self) -> String {
+        format_values!(self, |v: &Vec<T>| format!("[{}]", v.iter().join(", ")))
+    }
+}
+
+#[allow(clippy::redundant_closure_call)]
+impl<'a> ComparisonAssertionError<'a, U256> {
+    fn format_with_decimals(&self, decimals: &U256) -> String {
+        format_values!(self, |v: &U256| format_units_uint(v, decimals))
+    }
+}
+
+#[allow(clippy::redundant_closure_call)]
+impl<'a> ComparisonAssertionError<'a, I256> {
+    fn format_with_decimals(&self, decimals: &U256) -> String {
+        format_values!(self, |v: &I256| format_units_int(v, decimals))
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("Assertion failed: {left} !~= {right} (max delta: {max_delta}, real delta: {real_delta})")]
+struct EqAbsAssertionError<T, D> {
+    left: T,
+    right: T,
+    max_delta: D,
+    real_delta: D,
+}
+
+impl EqAbsAssertionError<U256, U256> {
+    fn format_with_decimals(&self, decimals: &U256) -> String {
+        format!(
+            "{} !~= {} (max delta: {}, real delta: {})",
+            format_units_uint(&self.left, decimals),
+            format_units_uint(&self.right, decimals),
+            format_units_uint(&self.max_delta, decimals),
+            format_units_uint(&self.real_delta, decimals),
+        )
+    }
+}
+
+impl EqAbsAssertionError<I256, U256> {
+    fn format_with_decimals(&self, decimals: &U256) -> String {
+        format!(
+            "{} !~= {} (max delta: {}, real delta: {})",
+            format_units_int(&self.left, decimals),
+            format_units_int(&self.right, decimals),
+            format_units_uint(&self.max_delta, decimals),
+            format_units_uint(&self.real_delta, decimals),
+        )
+    }
+}
+
+type ComparisonResult<'a, T> = Result<Vec<u8>, ComparisonAssertionError<'a, T>>;
+
+impl Cheatcode for assertTrue_0Call {
+    fn apply(&self, _state: &mut Cheatcodes) -> Result {
+        Ok(assert_true(self.condition).map_err(|_| "Assertion failed")?)
```

Comment:
This kind of conflicts with your other suggestion to rename error message, by default error on assert is "Assertion failed", and in that case it would be "assertion failed"

which one should I do?

Reply:
I don't see how they conflict, and error messages should be lowercase